agent/consul/acl_idp.go

```diff
@@ -0,0 +1,279 @@
+package consul
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+type IdentityProviderValidator interface {
+	// ValidateToken takes raw user-provided IdP metadata and ensures it is
+	// sane, provably correct, and currently valid. Relevant identifying data
+	// is extracted and returned for immediate use by the role binding process.
+	//
+	// Depending upon the provider, it may make sense to use these calls to
+	// continue to extend the life of the underlying token
+	//
+	// TODO(rb): cook up stock errors that the caller may care about
+	ValidateLogin(*LoginValidationRequest) (*LoginValidationResponse, error)
+
+	AvailableFields() []string
+}
+
+type LoginValidationRequest struct {
+	// Token is a bearer token, such as a JWT that is identity provider specific.
+	Token string
+}
+
+type LoginValidationResponse struct {
+	// Fields stores IdP specific metadata suitable for the Role Binding process.
+	Fields map[string]string
+}
+
+type idpValidatorEntry struct {
+	Validator   IdentityProviderValidator
+	ModifyIndex uint64 // the raft index when this last changed
+}
+
+func (s *Server) purgeIdentityProviderValidators() {
+	s.aclIDPValidatorLock.Lock()
+	s.aclIDPValidators = make(map[string]*idpValidatorEntry)
+	s.aclIDPValidatorLock.Unlock()
+}
+
+func (s *Server) getIdentityProviderValidator(name string) (uint64, IdentityProviderValidator, bool) {
+	s.aclIDPValidatorLock.RLock()
+	defer s.aclIDPValidatorLock.RUnlock()
+
+	if s.aclIDPValidators != nil {
+		v, ok := s.aclIDPValidators[name]
+		if ok {
+			return v.ModifyIndex, v.Validator, true
+		}
+	}
+	return 0, nil, false
+}
+
+func (s *Server) getOrReplaceIdentityProviderValidator(name string, idx uint64, v IdentityProviderValidator) IdentityProviderValidator {
+	s.aclIDPValidatorLock.Lock()
+	defer s.aclIDPValidatorLock.Unlock()
+
+	if s.aclIDPValidators == nil {
+		s.aclIDPValidators = make(map[string]*idpValidatorEntry)
+	}
+
+	prev, ok := s.aclIDPValidators[name]
+	if ok {
+		if prev.ModifyIndex >= idx {
+			return prev.Validator
+		}
+	}
+
+	s.logger.Printf("[INFO] acl: updating cached identity provider validator for %q", name)
+
+	s.aclIDPValidators[name] = &idpValidatorEntry{
+		Validator:   v,
+		ModifyIndex: idx,
+	}
+	return v
+}
+
+func isValidIdentityProviderField(idpType, name string) bool {
+	var allowed []string // this list will be VERY short
+
+	switch idpType {
+	case "kubernetes":
+		allowed = k8sAvailableFields
+	}
+
+	for _, f := range allowed {
+		if f == name {
+			return true
+		}
+	}
+
+	return false
+}
+
+func findUnknownIdentityProviderFields(idpType string, names []string) []string {
+	if len(names) == 0 {
+		return nil
+	}
+
+	var (
+		allowed []string // this list will be VERY short
+		unknown []string
+	)
+
+	switch idpType {
+	case "kubernetes":
+		allowed = k8sAvailableFields
+	}
```

Comment:
Is there any value in making things like this part of the IdP interface rather than having multiple helper functions that switch and would all need to be modified to add a new IdP?

I like the idea of IdPs being an abstraction that doesn't leak in too many places so that we can add new ones without adding new cases to switches across several files.

Not a big deal though as they are mostly contained in this file but it seems like it could be simpler to make that a bit cleaner.